### DISCUSSION TITLE
[Feature request] add option to print the DAG / order in which releases will be deployed

### DISCUSSION THREAD

**Discussion by luispabon** [helmfile/helmfile#1368] [Ideas]
**Note:** I'm carrying over this request from @derlin at the old repo here: https://github.com/roboll/helmfile/issues/1916 

--- 

When the helmfile is very large, it is difficult to get the needs right. We sometimes forget one, etc. However, there is no easy way to see the order that will be used during deployment (or the DAG). The only thing we can do is a `helmfile sync` or `helmfile apply`, which takes a lot of time and requires a Kubernetes project, and wait for potential errors.

Since there is already all the code to compute the DAG, It would be nice to make it available in a command. The idea would be to output either the dag itself, or simply the list of releases in the order of deployment.

For example:

```
# either as a list, and then we could reuse the list command
helmfile list --ordered
# or as a specific comment, with a tree-like output:
helmfile show dag
```

The environment flag should also be supported, with for example `default` as the default.

**Comment by luispabon** [comment]
Our specific use case:

We have a fairly large helmfile repository from which we generate and publish helm charts and value & argocd config files for our gitops CD system. Our helmfiles establish an order of installation of released based on `needs`, pretty standard stuff that works fine.

In order to implement our argo `ApplicationSets` with rolling syncs to install everything in a certain order we're leveraging helmfile's DAG, using [this workaround](https://blog.derlin.ch/helmfile-understand-and-visualize-the-order-in-which-releases-are-deployed) (note: you can use `helmfile template` with it, not just `sync`) which is hacky to say the least and could break at any helmfile update if the wording of the output ever changes.

It would be good to have a single command that would just print those release groups, perhaps as a json file?

The current debug output looks like:
```
processing 3 groups of releases in this order:
GROUP RELEASES
1     orwell, freud, bradburry, wells, weber 
2     foo, bar
3     buzz
```

Instead, for example:
```
~ helmfile list-dag
```
```json
[
    {
        "group": 1,
        "releases": [
            "orwell",
            "freud",
            "bradbury",
            "wells",
            "weber"
        ]
    },
    {
        "group": 2,
        "releases": [
            "foo",
            "bar"
        ]
    },
    {
        "group": 3,
        "releases": [
            "buzz"
        ]
    }
]


```

**Comment by luispabon** [comment]
Looks like there's prior art here:
https://github.com/justindriggers/helmfile/commit/19d433c3dad9dc93963f4a4bfd772cded8609e9f

**Comment by yxxhero** [comment]
@luispabon PR is welcome.

**Reply by luispabon** [reply]
I'll see what I can do, I haven't touched go for a few years.

**Comment by yxxhero** [comment]
https://github.com/helmfile/helmfile/pull/1477

**Comment by luispabon** [comment]
Amazing, thank you all
